spot_rl_experiments/spot_rl/skills/atomic_skills.py

```diff
@@ -33,49 +35,196 @@
+        self.spot = spot
+        self.config = config
+        self.verbose = True
+
+        # Log the most recent robot trajectory
+        self.robot_trajectory = []  # type: List[Dict]
+        self.start_time = None  # type: float
+
+        # Setup (Do Not create policy and env objects here, they should be created in the derived atomic classes)
+        # self.policy = None
+        # self.env = None
+
+    def reset_skill(self, goal_dict: Dict[str, Any]):
+        """
+        Performs:
+            Sanity checks the input goal_dict,
+            Resets the env and policy,
+            Updates the logged data at init,
+            Sets the start time for recording before starting skill
+
+        Args:
+            goal_dict: Dict[str, Any] containing necessary goal information for skill
+
+        Returns:
+            observations: Initial observations from the env
+        """
+        raise NotImplementedError(
+            "This method should be implemented in the derived class"
+        )
+
+    def execute_rl_loop(self, goal_dict: Dict[str, Any]) -> Tuple[bool, str]:
+        """
+        !!BLOCKING CALL!!
+
+        Executes the action based on output from policy
+
+        WARNING: This method WILL NOT perform any explicit sanity checks on the input provided.
+                 It is the responsibility of the caller to ensure that the input is valid
+                 and safe to use or perform sanity checks inside reset_skill() method
+
+        Args:
+            goal_dict: Dict[str, Any] containing necessary goal information for the appropriate skill
+
+        Returns:
+            status (bool): Whether robot was able to succesfully execute the skill or not
+            message (str): Message indicating description of success / failure reason
+        """
+        # Reset policy,env and update logged data at init
+        try:
+            observations = self.reset_skill(goal_dict)
+        except Exception as e:
+            raise e
+        done = False
+
+        # Execution Loop
+        while not done:
+            action = self.policy.act(observations)  # type: ignore
+            action_dict = self.split_action(action)
+            observations, _, done, _ = self.env.step(action_dict=action_dict)  # type: ignore
+
+            # Record trajectories at every step
+            self.robot_trajectory.append(
+                {
+                    "timestamp": time.time() - self.start_time,
+                    "pose": [
+                        self.env.x,  # type: ignore
+                        self.env.y,  # type: ignore
+                        np.rad2deg(self.env.yaw),  # type: ignore
+                    ],
+                }
+            )
+
+        # Update logged data after finishing execution and provide status of execution
+        return self.update_and_check_status(goal_dict)
+
+    def execute(self, goal_dict: Dict[str, Any]) -> Tuple[bool, str]:  # noqa
+        raise NotImplementedError(
```

Comment:
why is this implemented in derived class?

Reply:
Moved to base class for Nav & Pick. Keeping in derived class for Place because place.execute() also has extra code for handling IK based arm motion.